Extend readme with 'beginners guide'

bollu opened the pull request:
<!-- Thank you for your contribution.
     Make sure you read the contributing guide and fill this template. -->


<!-- Keep what applies -->
**Kind:** documentation

This is just some text added from discussions on `gitter` and questions I asked on #7804 about getting a sane development environment for Coq up and running.

I heavily edited the docs. Here are some useful links to the new docs:
- [`CONTRIBUTING.md` edited pull requests section](https://github.com/bollu/coq/blob/jun-27-add-stuff-i-learnt-to-beginners-guide/CONTRIBUTING.md#pull-requests)
- [`dev/`](https://github.com/bollu/coq/tree/jun-27-add-stuff-i-learnt-to-beginners-guide/dev)
- [`dev/doc/README.md](https://github.com/bollu/coq/blob/jun-27-add-stuff-i-learnt-to-beginners-guide/dev/doc/README.md)
- [`dev/doc/emacs.md`](https://github.com/bollu/coq/blob/jun-27-add-stuff-i-learnt-to-beginners-guide/dev/doc/emacs.md)
- [`dev/doc/vim.md`](https://github.com/bollu/coq/blob/jun-27-add-stuff-i-learnt-to-beginners-guide/dev/doc/vim.md)
- [`README.md`](https://github.com/bollu/coq/blob/jun-27-add-stuff-i-learnt-to-beginners-guide/README.md) Possibly controversial change: I pulled `Contributing` up so it catches the eye more easily :) Feel free to reject this.

I stole some content off of the wiki as well. I wish to delete the wiki pages, since they only seem to contain some info on vim / emacs setup, and not much else. [Wiki page on `DevelSetup`](https://github.com/coq/coq/wiki/DevelSetup). I am not sure how useful the `git` section of `DevelSetup` is. My vote would be to not have it, since we point people to Github's tutorial of sending PR's anyway.

Some of the old `setup.txt` we into `dev/doc/README.md`, other parts of it into `dev/doc/emacs.md`

bollu commented:
Also, may I update the file to `.md` for nicer looking formatting on github?

SkySkimmer commented on dev/README:
```diff
@@ -2,6 +2,38 @@ This directory contains information and tools to help develop the
+the correct `configure` settings, then hacking on Coq, make-ing, and testing.
+
+First of all, it is recommended to run the configure with
+    ./configure -local -devel
```
Isn't it `-profile devel`?

Zimmi48 replied:
Indeed, and this subsumes `-local`.

SkySkimmer commented on dev/README:
```diff
@@ -2,6 +2,38 @@ This directory contains information and tools to help develop the
+which build bytecode interpreter versions of coqtop.
+
+To access coqc from coqtop.byte, run:
+    coqtop.byte -compile
```
Interesting targets:
- `states` builds just enough to run coqtop (including the .vo files which are automatically loaded)
- `byte` for building all the bytecode stuff

I don't know why you would run coqtop.byte over coqtop outside ocamldebug.

SkySkimmer commented on dev/README:
```diff
@@ -2,6 +2,38 @@ This directory contains information and tools to help develop the
+So, a typical setup might be:
+    $ ./configure -local -devel
+    <hack>
+    $ make -j bin/coqtop.byte pluginsbyte
```
Don't use `-j` without argument.

SkySkimmer commented on dev/README:
```diff
@@ -2,6 +2,38 @@ This directory contains information and tools to help develop the
+
+
+To run the test suite, run `make` from the `test-suite` directory.
+Running `make -B` will always rerun all the tests (`-B` is to remake everything)
```
Just `make clean` before rerunning whatever.

SkySkimmer commented on dev/README:
```diff
@@ -2,6 +2,38 @@ This directory contains information and tools to help develop the
+Running `make -B` will always rerun all the tests (`-B` is to remake everything)
+This is explained in greater detail in 
+[`test-suite/README.md`](https://github.com/coq/coq/blob/master/test-suite/README.md)
+    $ make -C -j test-suite/
```
Don't use `-j` without argument.

SkySkimmer commented on dev/README:
```diff
@@ -2,6 +2,38 @@ This directory contains information and tools to help develop the
+First of all, it is recommended to run the configure with
+    ./configure -local -devel
+
+-local allows `make install` to create a local dev install.
```
This is incorrect, rather `-local` means there is no install step, Coq is used from the build directory.

SkySkimmer commented on dev/README:
```diff
@@ -2,6 +2,38 @@ This directory contains information and tools to help develop the
+
+So, a typical setup might be:
+    $ ./configure -local -devel
+    <hack>
```
Usually one would build before hacking so that merlin works.

bollu commented:
@Zimmi48 In that case, may I:

1. Modernize (i.e, change to markdown `dev/README` to `dev/README.md` that adds links to all files.
2. add a `dev/doc/README.md` that adds "developer documentation" and fuses all of the info together
3. Point to `dev/doc/README.md` and `dev/README.md` in the `CONTRIBUTING.md` file?

I still think this is too much information spread across too many files, though, but I think that is unavoidable considering the scale of Coq :)

Zimmi48 commented:
Yes :sparkles:, if you could do this it would be awesome. This is the kind of things I had wanted to do for a long time but couldn't just because I had so much else on my plate :confused:

bollu commented:
Ping @Zimmi48 , @SkySkimmer . I heavily edited some of the files and added links. Please have a look.
I also updated the PR text pointing out the changes.

SkySkimmer commented on dev/doc/emacs.md:
```diff
@@ -0,0 +1,233 @@
+# Emacs configuration for Coq development
```
Isn't this stuff from the wiki?

bollu replied:
Yes, which is why I want to remove the wiki page (there's a comment at the bottom of the file that says `delete wiki page once merged`). This was a general sweep to consolidate information.

SkySkimmer replied:
Somehow I like it better on the wiki. Do we have any guidelines for deciding what goes in the wiki and what goes in the repo?

bollu replied:
None that I know.

If you like it better on the wiki, then I propose we move whatever I've put into `emacs.md` and `vim.md` into the wiki.

I wanted it in the repo so it is easier to link to (and to grep for)

Zimmi48 replied:
I think I agree with putting the editor configs on the wiki rather than on the repo since it's likely to change independently of Coq.

Zimmi48 commented on CONTRIBUTING.md:
```diff
@@ -20,6 +20,9 @@ If you want to minimize your bug (or help minimize someone else's) for more extr
 ## Pull requests
 
+##### Beginner's guide: [`dev/doc/README.md`](dev/doc/README.md)
+##### Development information and tools: [`dev/README.md`](dev/README.md)
```
Not a fan of abusing titles, can you use strong emphasis `** **` instead?

Zimmi48 commented on README.md:
```diff
@@ -27,6 +27,10 @@ See also [Cocorico](https://github.com/coq/coq/wiki) (the Coq wiki),
+## Contributing
+
+Guidelines for contributing to Coq in various ways are listed in the [contributor's guide](CONTRIBUTING.md).
+
```
Indeed, I don't think I like this change. I'm unsure this makes it easier to find "Contributing", and it moves it away from "Bug reports" and in between two sections that are both about documentation. What I would agree with would be to reduce the number of main sections (for instance by making Changes a sub-section of Documentation, and Bug reports a sub-section of Contributing).

Also, note that now the contributing guide is linked to from the main page of the Coq website https://coq.inria.fr/

Zimmi48 commented on dev/README.md:
```diff
@@ -0,0 +1,41 @@
+
+
+## Debugging and profiling (`dev/`)
+##### More info: [`doc/debugging.md`](doc/debugging.md)
```
Same comment on abusing titles.

Zimmi48 commented on dev/README.md:
```diff
@@ -0,0 +1,41 @@
+| [`dev/doc/style.txt`](doc/style.txt) | a few style recommendations for writing Coq ML files |
+| [`dev/doc/debugging.md`](/doc/debugging.md) | help for debugging or profiling |
+| [`dev/doc/universes.txt`](/doc/universes.txt) |  help for debugging universes |
+| [`dev/doc/translate.txt`](/doc/translate.txt) | help for using coq translator |
```
This file is probably outdated, and could be archived / removed.

bollu replied:
@Zimmi48 which file, translate? The file is still there, though? Do you want me to **remove** `dev/doc/translate.txt` from the repo, as well, as remove the reference?

Zimmi48 replied:
Yes. It's probably time to remove this file.

Zimmi48 commented on dev/doc/README.md:
```diff
@@ -0,0 +1,90 @@
+
+$ source ~/.bashrc
+
+# needed if you want to build "coqtop" target
```
camlp5 is basically a strong dependency (needed all the time)

Zimmi48 commented on dev/doc/README.md:
```diff
@@ -0,0 +1,90 @@
+# needed if you want to build "doc" target
+$ sudo apt-get install texlive-latex-recommended \
+  texlive-fonts-extra texlive-math-extra \
+  hevea texlive-latex-extra latex-xcolor
```
This info is outdated. It's probably better to just link to `doc/README.md` to get this information, isn't it?

bollu replied:
Possibly, this was stuff that was in `dev/doc/install.txt` that I simply retained.

Zimmi48 commented on dev/doc/README.md:
```diff
@@ -0,0 +1,90 @@
+$ bin/coqtop -compile <test_file_name.v>
+<goto hack until stuff works>
+
+$ make -C test-suite -j $JOBS # run test suite
```
If the intent is to run the full test-suite, `make test-suite` works too. `make -C test-suite` is useful to only re-run failed jobs. I guess this "advanced" info can be left out to test-suite/README.md

Zimmi48 commented on dev/doc/README.md:
```diff
@@ -0,0 +1,90 @@
+
+
+To run the test suite, run `make` from the `test-suite` directory.
+This is explained in greater detail in [`/test-suite/README.md`](/test-suite/README.md)
```
I recently discovered that relative links were better, see #7800. So you can use `../../test-suite/README.md` instead.

Zimmi48 commented on dev/doc/emacs.md:
```diff
@@ -0,0 +1,233 @@
+
+### TODO: Siddhart's strong opinion on this!!!
+I feel uncomfortable that Coq maintains docs to this detail about editor details.
+I would prefer it we had a short section on company and we then link to company docs. Thoughts?
```
I agree. Or at least, this should be in the wiki as an "unofficial resource".

bollu commented:
I moved the emacs and vim docs into the wiki: https://github.com/coq/coq/wiki/DevelSetup. I'm going to remove the emacs and vim files from the repo and point the `dev/README.md` to the wiki.

bollu commented:
@Zimmi48 OK, I edited files as requested, [and moved the dev pages to the wiki](https://github.com/coq/coq/wiki/DevelSetup). Can I have another round of review, please?

Zimmi48 commented:
BTW, I'll ask you to rewrite your commit history a bit before merging. I don't have any specific expectations on the number or the size of commits but don't leave commits with "WIP" in the commit message and try to have commit messages starting with a capital letter if possible. When you do this, you can probably rebase at the same time to have a greener CI (even if we don't really care about CI for this PR but green always feels nicer ^^).

bollu commented:
@Zimmi48 merged into one commit, cleaned up the new comments, rebased, and pushed.

Zimmi48 commented on dev/doc/README.md:
````diff
@@ -0,0 +1,82 @@
+```
+$ cd test-suite
+$ make clean && make -j $JOBS
+```
````
Sorry I wasn't clear earlier, but I think you can remove these two lines.

Zimmi48 commented on dev/doc/README.md:
```diff
@@ -0,0 +1,82 @@
+- `Coqloop.loop` : This function implements the read-eval-print loop.
+- `Vernacentries.interp` : This function is called to execute the Vernacular command user have typed.
+                       It dispatches the control to specific functions handling different Vernacular command.
+- `Vernacentries.vernac_check_may_eval` : This function handles the `Check ...` command.
```
Nitpicking: you could remove the spaces before the colons.

Zimmi48 commented:
@bollu It looks like you have done something weird while rebasing.

bollu commented:
@Zimmi48 fixed, I think

Zimmi48 commented:
Yes, but it looks like you only fixed one out of the two last nits...

bollu commented:
@Zimmi48 whoops, fixed now..
